espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/dispatcher/KeyDispatch.java

```diff
@@ -0,0 +1,263 @@
+package io.appium.espressoserver.lib.helpers.w3c.dispatcher;
+
+import javax.annotation.Nullable;
+
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
+import io.appium.espressoserver.lib.helpers.w3c.adapter.W3CActionAdapter;
+import io.appium.espressoserver.lib.helpers.w3c.dispatcher.constants.KeyCodeMapper;
+import io.appium.espressoserver.lib.helpers.w3c.dispatcher.constants.KeyLocationMapper;
+import io.appium.espressoserver.lib.helpers.w3c.dispatcher.constants.KeyNormalizer;
+import io.appium.espressoserver.lib.helpers.w3c.dispatcher.constants.NormalizedKeys;
+import io.appium.espressoserver.lib.helpers.w3c.models.ActionObject;
+import io.appium.espressoserver.lib.helpers.w3c.state.KeyInputState;
+
+/**
+ * Implement key dispatch events (https://www.w3.org/TR/webdriver/#keyboard-actions)
+ */
+public class KeyDispatch {
+
+    /**
+     * Dispatch a key down event (in accordance with https://www.w3.org/TR/webdriver/#keyboard-actions)
+     *
+     * @param dispatcherAdapter Adapter that has implementation specific keyDown event
+     * @param actionObject W3C Action Object
+     * @param inputState State of the input source
+     * @param tickDuration How long the 'tick' is. This is unused currently but may be used in the future
+     * @return KeyEvent (for testing purposes). 'null' if the dispatch failed.
+     */
+    @Nullable
+    public static KeyEvent dispatchKeyDown(W3CActionAdapter dispatcherAdapter,
+                                   ActionObject actionObject, KeyInputState inputState, long tickDuration) throws AppiumException {
+
+        // Get the base Key Event
+        KeyEvent keyEvent = getKeyEvent(actionObject);
+        String key = keyEvent.getKey();
+
+        // 3. If the input state's pressed property contains key, let repeat be true, otherwise let repeat be false.
+        boolean repeat = inputState.isPressed(key);
+
+        // 6. Let charCode, keyCode and which be the implementation-specific values of the charCode,
+        // keyCode and which properties.
+        // Omiting this for now until there's an implementation that needs it.
+
+        // 7-10: Set the alt, shift, ctrl meta key state
+        boolean alt = key.equals(NormalizedKeys.ALT);
+        boolean shift = key.equals(NormalizedKeys.SHIFT);
+        boolean ctrl = key.equals(NormalizedKeys.CONTROL);
+        boolean meta = key.equals(NormalizedKeys.META);
+        if (alt) inputState.setAlt(true);
```

Comment:
`inputState.setAlt(key.equals(NormalizedKeys.ALT))`
...

Reply:
Okay, but it has to be `inputState.setAlt(inputState.isAlt() || key.equals(NormalizedKeys.ALT));` because if `inputState.isAlt()` is _already_ true it needs to stay true.